src/event.c

```diff
@@ -267,10 +278,23 @@ subject_attr_t *get_subj_attr(event_t *e, subject_type_t t)
 		return sn;
 	}
 
-	free(subj.str);
+	// free .str only when it was really used
+	// otherwise invalid free is possible
+	if (t >= COMM) free(subj.str);
```

Comment:
Do we really need to free the string? It should live in the cache until the entry gets evicted and then it gets cleaned up. While testing, I got a segfault one time. I have not been able to recreate it with address sanitizer compiled in but this would be my best guess as to what caused it.

Reply:
Looked at this in more detail. I see what you are doing. The subj.str is only valid when subject type is COMM or higher. However, around line 213 subj.str was set to NULL. So, if the type were < COMM, str would still be NULL and it is legal to call free with NULL (just becomes a no-op). So, the code is safe without this change.

Reply:
Or, you could just do:

```
if (subj.str)
        free(subj.str);
```
I think that requires less head scratching and less ambiguity...

Reply:
> Or, you could just do:
> 
> ```
> if (subj.str)
>         free(subj.str);
> ```
> 
> I think that requires less head scratching and less ambiguous...

I think this is not enough because when ```val``` is set to 1 then ```str``` is very likely going to be something else then 0 and free() will be called on something that comes from bad interpretation.

Reply:
> Looked at this in more detail. I see what you are doing. The subj.str is only valid when subject type is COMM or higher. However, around line 213 subj.str was set to NULL. So, if the type were < COMM, str would still be NULL and it is legal to call free with NULL (just becomes a no-op). So, the code is safe without this change.

I was thinking that it can be better to check this and make this explicit rather than implicit. 
And BTW it works like that in ```subject_clear``` and ```subject_reset```.

Reply:
Huh? My understanding is that `subj.str` is either set or it's not. So, either NULL or !NULL, In NULL cases, do nothing as there is no resources to free anyway. For non-NULL cases, irrespective of the value, we free the any allocated resources...

Anyway, I could be wrong, but it *should* be this simple.

Reply:
> Huh? My understanding is that `subj.str` is either set or it's not. So, either NULL or !NULL, In NULL cases, do nothing as there is no resources to free anyway. For non-NULL cases, irrespective of the value, we free the any allocated resources...
> 
> Anyway, I could be wrong, but it _should_ be this simple.

To be on the same page, ```str``` and ```val``` are part of the union.
```
typedef struct s {
	subject_type_t type;
	union {
		int val;
		char *str;
	};
} subject_attr_t;
```
Therefore modifying ```val``` means also that NULL test on ```str``` is irrelevant.

Reply:
Ah I see. Fair enough.